### PR TITLE
[#187] README: Explain how to pass values back to the caller, across REP boundaries (main)

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -439,3 +439,46 @@ callback.writeLine('serverLog', 'hello, server log!')
...
+    # Now, we invoke the Python rule in core.py. Remember, the rule takes one
+    # input argument. That argument maps to index 0 within the python rule.
+    get_magic_number(*magic_number);
+
+    # Finally, the value is written to the log file.
+    writeLine('serverLog', 'The magic number is [*magic_number].');
+}
+```
+
+It's important the value passed across rule engine plugin boundaries be of type string. Passing any other type will result in an error.
```
Comment (alanking) [inline README.md:484]:
```suggestion
It's important that the value passed across rule engine plugin boundaries be of type string. Passing any other type will result in an error.
```

**Review by alanking — COMMENTED** [review]
Seems good

File: README.md [hunk 2]
```diff
@@ -439,3 +439,46 @@ callback.writeLine('serverLog', 'hello, server log!')
...
+
+This is achieved by writing a value back _(as a string)_ to the appropriate index within the `rule_args` array. The caller is required to pass one or more variables as **out** parameters.
+
+Here's an example demonstrating how to pass a value, via the input arguments, from the PREP to the native rule engine plugin (NREP).
+
+The following is a python function which overwrites the value of the first input argument.
+```python
+# file: core.py
+
+def get_magic_number(rule_args, callback.rei):
```
Comment (trel) [inline README.md:453]: is `callback.rei` correct?

Comment (korydraughn) [reply README.md:453]: Oh no. Will fix.

File: README.md (hunk 1 above)
Comment (korydraughn) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (korydraughn) [reply README.md]: done

**Comment by korydraughn** [comment]
Squashed.

Any last comments before I add the pound?

File: README.md [hunk 3]
```diff
@@ -439,3 +439,46 @@ callback.writeLine('serverLog', 'hello, server log!')
...
+
+And here is the NREP rule which invokes the PREP rule.
+```python
+# file: core.re
+
+nrep_rule()
+{
+    # Due to how the NREP works, we must initialize the variable to a string.
+    # The value of the string isn't important. The important part is that the
+    # variable be initialized to a string.
```
Comment (trel) [inline README.md:472]: the last sentence feels redundant.

File: README.md [hunk 4]
```diff
@@ -439,3 +439,46 @@ callback.writeLine('serverLog', 'hello, server log!')
...
+    *magic_number = '-1';
+
+    # Now, we invoke the Python rule in core.py. Remember, the rule takes one
+    # input argument. That argument maps to index 0 within the python rule.
+    get_magic_number(*magic_number);
+
+    # Finally, the value is written to the log file.
+    writeLine('serverLog', 'The magic number is [*magic_number].');
```
Comment (trel) [inline README.md:480]: could show after this code block the expected output (aka, show the 42 coming out).

File: README.md (hunk 3 above)
Comment (korydraughn) [reply README.md]: Removed it.

File: README.md (hunk 4 above)
Comment (korydraughn) [reply README.md:479]: I added a few more words to the bottom of the page about this. I may tweak the words more.

I'll need to compile the PREP to make sure things work as expected.

Comment (korydraughn) [reply README.md:479]: Done.

**Review by alanking — COMMENTED** [review]
Seems good to me. Squash em.

**Review by alanking — APPROVED** [review]
\# it

File: README.md [hunk 5]
```diff
@@ -439,3 +439,66 @@ callback.writeLine('serverLog', 'hello, server log!')
...
+}
+```
+
+It's important that the value passed across rule engine plugin boundaries be of type string. Passing any other type will result in an error.
+
+Upon invoking `nrep_rule`, a message very similar to the one that follows will appear in the server log file. See the **log_message** property. It should contain `[42]` instead of `[*magic_number]`.
```
Comment (trel) [inline README.md:485]: instead of `[-1]`?

Comment (korydraughn) [reply README.md:485]: -1 is overwritten with 42 by the python rule.

**Comment by korydraughn** [comment]
#'d

File: README.md (hunk 5 above)
Comment (korydraughn) [reply README.md:485]: I'll reword that.

How about ... `[*magic_number] was explanded to [42].`

Comment (trel) [reply README.md:485]: I think this is the clearest...

```
It should contain `[42]` instead of `[-1]`.
```

Comment (korydraughn) [reply README.md]: Done.

**Review by alanking — APPROVED** [review]

**Pull request merged**